Add support for IOS vlan parsing filter.

idahood opened the pull request:
##### SUMMARY
<!--- Describe the change, including rationale and design decisions -->
Add support for IOS vlan parsing filter.

<!--- If you are fixing an existing issue, please include "Fixes #nnn" in your
commit message and your description; but you should still explain what
the change does.-->

##### ISSUE TYPE
<!--- Pick one below and delete the rest: -->
 - Feature Pull Request

##### COMPONENT NAME
<!--- Name of the module, plugin, module or task -->
filter_plugin

##### ANSIBLE VERSION
<!--- Paste verbatim output from "ansible --version" between quotes below -->
```
ansible 2.5.3
  config file = /etc/ansible/ansible.cfg
  configured module search path = [u'/Users/sdodd/.ansible/plugins/modules', u'/usr/share/ansible/plugins/modules']
  ansible python module location = /usr/local/lib/python2.7/site-packages/ansible
  executable location = /usr/local/bin/ansible
  python version = 2.7.14 (default, Jan  6 2018, 12:16:16) [GCC 4.2.1 Compatible Apple LLVM 9.0.0 (clang-900.0.39.2)]
```


##### ADDITIONAL INFORMATION
<!--- Include additional information to help people understand the change here.
For bugs that don't have a linked bug report, a step-by-step reproduction
of the problem is helpful. -->

<!--- Paste verbatim command output below, e.g. before and after your change -->
```
Example Jinja template below:

{% set parsed_vlans = vlans | vlan_parser %}
switchport trunk allowed vlan {{ parsed_vlans[0] }}
{% for i in range (1, parsed_vlans | count) %}
switchport trunk allowed vlan add {{ parsed_vlans[i] }}

This allows for dynamic generation of vlan lists on a Cisco IOS tagged interface. You can store an exhaustive raw list of the exact vlans required for an interface and then compare that to the parsed IOS output that would actually be generated for the config. 
```

steve-dodd committed:
Add support for IOS vlan parsing filter.
Example usage below:

{% set parsed_vlans = vlans | vlan_parser %}
switchport trunk allowed vlan {{ parsed_vlans[0] }}
{% for i in range (1, parsed_vlans | count) %}
switchport trunk allowed vlan add {{ parsed_vlans[i] }}

idahood committed:
Update test_network.py

Add import statement for filter

steve-dodd committed:
Fixed PEP8 issues relating to comments

steve-dodd committed:
Fix PEP8 issues related to blank lines

idahood commented:
bot_status

idahood commented:
ready_for_review

itdependsnetworks commented on lib/ansible/plugins/filter/network.py:
```diff
@@ -395,6 +395,79 @@ def comp_type5(unencrypted_password, encrypted_password, return_orginal=False):
+    for vlans in parse_list:
+        # First line (" switchport trunk allowed vlan ")
+        if line_count == 0:
+            if len(result[line_count] + vlans) > 48:
```
I think it would be great if this was configurable, and perhaps defaulted to this value? I suspect other similar OS's have slightly different numbers.

idahood replied:
@itdependsnetworks what aspect do you want configurable? This particular logic is built around IOS vlan lists specifically, I can't really comment if it applies to JunOS etc. Happy to rename to something besides 'vlan_parser' to indicate as such. 'ios_style_vlans' ?

itdependsnetworks replied:
"what aspect do you want configurable": '{{set parsed_vlans = vlans | vlan_parser(first_line_len=48, other_line_count=40) }}`

Non-ios, but ios like devices such as eos, nxos, etc.. They all have similar structure. Additionally, it may work fine for your version, but another version may have 50 char's.

idahood replied:
I have incorporated the requested change.

steve-dodd committed:
Removed magic numbers for line lengths. This should generalize support
to other IOS-like NOS that use similar methods for listing vlans. The
default arguments for line lengths will still be specific to Cisco IOS.
The unit tests for line length are still specific to Cisco IOS.

idahood commented:
ready_for_review

ryanmerolle reviewed (approved):
Nice job.  Very useful, and it works as described.
